### DISCUSSION TITLE
What is the correct way to stop AsyncClient ?

### DISCUSSION THREAD

**Discussion by bialix** [miguelgrinberg/python-socketio#927] [Q&A]
I can't figure how to properly shut down AsyncClient on Ctrl+C or SIGINT or something similar. Everything I've tried with no luck.
I always get this warning with Python 3.6:

```
D:\work\Opensource\socketio-example\async-client\venv\lib\site-packages\engineio\asyncio_client.py:206:\
   RuntimeWarning: coroutine 'ClientSession.close' was never awaited
ERROR - asyncio - Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x000000FF462BC588>
```

My code for client is endless loop because I need to use AsyncClient with other asyncio code which do other operations. So I extracted the code related to socketio into working example and publish it here:

https://github.com/bialix/socketio-async-client-example

There are 2 files: client-example.py and client-example2.py. The former is simple code, the latter trying to catch Ctrl+C signal and do graceful shutdown based on the examples I found on the internet. No matter what I've tried I can't get rid of aiohttp.client.ClientSession warning.

I need help, please.

**Comment by bialix** [comment]
I've extracted the simplified example code here:

```
class Client:
    def __init__(self):
        self.stop = False
        self.sio = socketio.AsyncClient()

    async def main(self):
        while not self.stop:
            try:
                await self.sio.connect("http://127.0.0.1:10069")
                await self.sio.wait()
            except socketio.exceptions.ConnectionError:
                await asyncio.sleep(0.1)

def main():
    k = Client()
    loop = asyncio.get_event_loop()
    try:
        group = asyncio.gather(k.main(), other_async_code)
        loop.run_until_complete(group)
    finally:
        loop.close()

if __name__ == "__main__":
    sys.exit(main())
```

**Reply by miguelgrinberg** [reply]
I'm not sure this is possible. The signal handling for asyncio is not very reliable. I'm definitely not going to introduce fixes for 3.6, which I'm not testing for anymore, but if the problem is still present in 3.10 I can give this another look.

**Comment by bialix** [comment]
I've just tested with Python 3.10 on Windows. The same behaviour. Hit Ctrl+C and got the same warning.
Also, ``Signal handler is unsupported`` warning shown up but it should not be here.

**Comment by bialix** [comment]
Ok, it seems my client-example2.py does actually works as expected on Python 3.10. Ok.

UPD: no, that's not quite correct. The clean exit is possible only when connect is established. But if we waiting for connect then the same warning happens.

I suppose it's a leaky astractions.

**Comment by bialix** [comment]
Just to add more info, on Linux with Python 3.6 I've got the same problem but traceback is different:

```
[bialix@rpmbuild-dev socketio-async-client-example]$ ./python client-example2.py
DEBUG - main - attach signal SIGABRT
DEBUG - main - attach signal SIGTERM
DEBUG - main - attach signal SIGINT
DEBUG - main - trying to connect
DEBUG - main - connection established
DEBUG - main - send > {'foo': 'bar'}
^CDEBUG - main - disconnected from server
DEBUG - main - stopped
Traceback (most recent call last):
  File "client-example2.py", line 146, in <module>
    sys.exit(main())
  File "client-example2.py", line 95, in main
    loop.run_until_complete(group)
  File "/usr/lib64/python3.6/asyncio/base_events.py", line 482, in run_until_complete
    raise RuntimeError('Event loop stopped before Future completed.')
RuntimeError: Event loop stopped before Future completed.
/home/bialix/.local/lib/python3.6/site-packages/engineio/asyncio_client.py:206: RuntimeWarning: coroutine 'ClientSession.close' was never awaited
ERROR - asyncio - Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x7fb7320d9748>
```

My signal handlers are ignored, AsyncClient is disconnected though, but underlying ``aiohttp.client.ClientSession`` is still not happy.

Unfortunately I don't have access to Python 3.10 there. It's an old CentOS 7.

UPD: the same behaviour with Python 3.8 @ Ubuntu 20.04.4

**Comment by bialix** [comment]
It seems odd that my signal handlers are ignored on Linux. Probably due the fact underlying engineio registers their own asyncio.loop.add_signal_handler?

**Comment by bialix** [comment]
Oh, my. I think I found the bug in your ``engineio.async_client``.

Method ``async def disconnect(self, abort=False):`` executes ``await self.http.close()`` **only if** object is connected. Which is clearly wrong.
Simple doing this regardless of the ``connected`` state fixes the problem:

```
        if self.http and not self.http.closed:  # pragma: no cover
            await self.http.close()
```

Also, I have to execute at the start of my app:

```
    import engineio.asyncio_client
    engineio.asyncio_client.async_signal_handler_set = True
```

To be able to attach my own signals and do my own proper shutdown.

**Comment by bialix** [comment]
Filed bug report: https://github.com/miguelgrinberg/python-engineio/issues/277
